compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,211 @@
+        return buildFun {
+            name = function.name
+            returnType = function.returnType.substitute(typeMap)
+            visibility = function.visibility
+            origin = IrDeclarationOrigin.IR_BUILTINS_STUB
+            modality = Modality.OPEN
+        }.apply {
+            // Replace Function metadata with the data from class
+            // Add the abstract function symbol to stub function for bridge lowering
+            overriddenSymbols.add(function.symbol)
+            parent = irClass
+            dispatchReceiverParameter = function.dispatchReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            extensionReceiverParameter = function.extensionReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            function.valueParameters.mapIndexed { i, param ->
+                valueParameters.add(i, param.copyWithTypeErasure(this, typeMap))
+            }
+            val exception = context.getTopLevelClass(FqName("java.lang.UnsupportedOperationException"))
+            // Function body consist only of throwing UnsupportedOperationException statement
+            body = context.createIrBuilder(function.symbol).irBlockBody {
+                +irThrow(
+                    irCall(
+                        exception.owner.constructors.single {
+                            it.valueParameters.size == 1 && it.valueParameters.single().type.isNullableString()
+                        }
+                    ).apply {
+                        putValueArgument(0, irString("Operation is not supported for read-only collection"))
+                    }
+                )
+            }
+        }
+    }
+
+    // Type erasure for value parameters, with type substitution
+    private fun IrValueParameter.copyWithTypeErasure(target: IrSimpleFunction, typeMap: Map<IrTypeParameterSymbol, IrType>)
+            : IrValueParameter {
+        val descriptor = WrappedValueParameterDescriptor(this.descriptor.annotations)
+        return IrValueParameterImpl(
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            IrDeclarationOrigin.IR_BUILTINS_STUB,
+            IrValueParameterSymbolImpl(descriptor),
+            name,
+            index,
+            type.substitute(typeMap).eraseTypeParameters(),
+            varargElementType?.substitute(typeMap)?.eraseTypeParameters(),
+            isCrossinline,
+            isNoinline
+        ).apply {
+            descriptor.bind(this)
+            parent = target
+        }
+    }
+
+    // Create a map for type parameters between source class (Mutable Collection classes) to
+    // target class (class currently in lowering phase), this map is later used for substituting type parameters in generated functions
+    private fun createTypeMap(readOnlyClass: IrClass, mutableClass: IrClass, targetClass: IrClass): Map<IrTypeParameterSymbol, IrType> {
```

Comment:
I propose to use the existing term from our compiler frontend and rename this function to `computeSubstitutionMap` and adapt the comment accordingly